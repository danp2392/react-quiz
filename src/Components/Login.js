import React from "react";
import { useAuth } from "../auth/auth";
import { history } from "../routes/AppRouter";

import logo192 from "../images/logo192.png";

import Button from "@material-ui/core/Button";
import Box from "@material-ui/core/Box";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core";
export const Login = () => {
  const { startLogin } = useAuth();

  const login = async () => {
    try {
      await startLogin();

      history.push("/welcome");
    } catch (e) {
      throw new Error(e);
    }
  };

  const useStyles = makeStyles((theme) => ({
    loginDiv: {
      paddingTop: "20px",
      paddingBottom: "20px",
    },

    login: {
      lineHeight: "5rem",
      margin: "20px",
      color: "white",
      backgroundColor: theme.palette.primary.dark,
      fontSize: "2.5rem",
      "&:hover": {
        backgroundColor: "#8b9fa7",
      },
    },
  }));

  const classes = useStyles();

  return (
    <Box className={classes.loginDiv}>
      <img src={logo192} alt="react-logo" className={classes.reactLogo} />

      <Box className={classes.welcomeText}>
        <Typography variant="h3">React Quiz App</Typography>
        <Typography variant="h5">
          Take 10 random questions and find out how smart you are
        </Typography>
      </Box>
      <Button className={classes.login} onClick={login}>
        LOGIN
      </Button>
    </Box>
  );
};
