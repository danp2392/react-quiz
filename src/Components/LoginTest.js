import React from "react";
import { useAuth } from "../auth/auth";
import { history } from "../routes/AppRouter";
import logo192 from "../images/logo192.png";
import Button from "@material-ui/core/Button";
import Box from "@material-ui/core/Box";
import Typography from "@material-ui/core/Typography";
import { useStyles } from "../styles/LoginStyle";

export const LoginTest2 = () => {
  const classes = useStyles();
  return (
    <Box className={classes.loginDiv}>
      <img src={logo192} alt="react-logo" />

      <Box className={classes.welcomeText}>
        <Typography variant="h3">React Quiz App</Typography>
        <Typography variant="h5">
          Take 10 random questions and find out how smart you are
        </Typography>
      </Box>
      <Button data-testid="login-button">LOGIN</Button>
    </Box>
  );
};
