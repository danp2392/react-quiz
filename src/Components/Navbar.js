import React from "react";
import { NavLink } from "react-router-dom";
import { useAuth } from "../auth/auth";
import "../App.css";
import "../styling/Navbar.css";
import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
export const Navbar = () => {
  const { logOut } = useAuth();
  return (
    <Container>
      <AppBar className="topnav">
        <Toolbar>
          <NavLink className="nav-link" to="/welcome">
            <Typography> Home</Typography>
          </NavLink>

          <NavLink className="nav-link" to="/highscores">
            <Typography> High Scores</Typography>
          </NavLink>
          <NavLink className="nav-link" to="/" onClick={logOut}>
            <Typography> Logout</Typography>
          </NavLink>
        </Toolbar>
      </AppBar>
    </Container>
  );
};
