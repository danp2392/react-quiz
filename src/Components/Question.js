import React from "react";
import { decodeHTML } from "../functions/decodeHTML";
import "../App.css";
import "../styling/quiz.css";

export const Question = ({ id, question }) => {
  return (
    <div class="container">
      <div id="question-div">{question}</div>
    </div>
  );
};

export const Answer = ({ answer, id, checkAnswer }) => {
  return (
    <div class="container">
      <div id="answer-div">
        <div onClick={checkAnswer} id={id} class="answer">
          <li>{decodeHTML(answer)}</li>
        </div>
      </div>
    </div>
  );
};

export const ListAnswer = ({ answers, checkAnswer }) => {
  //to stop array from been empty, do some error checking
  //must have answers and must have a length and then map it
  //then it won't error out on click
 

  const answersList =
    answers &&
    answers.length &&
    answers.map((answer) => (
      <div>
        <Answer key={answer} answer={answer} checkAnswer={checkAnswer} />
      </div>
    ));

  return <ul id="list-div">{answersList}</ul>

};
