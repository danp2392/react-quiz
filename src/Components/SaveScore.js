import { db } from "../auth/firebase";
import React from "react";
import firebase from "firebase";
import { history } from "../routes/AppRouter";
import { useAuth } from "../auth/auth";
import Button from "@material-ui/core/Button";
import { Typography } from "@material-ui/core";
import { useStyles } from "../styles/SaveScoreStyle";

export const SaveScore = ({ score, name }) => {
  const timestamp = firebase.firestore.FieldValue.serverTimestamp();

  const { currentUser } = useAuth();
  const Save = () => {
    try {
      db.collection("users").doc(currentUser.uid).collection("scores").add({
        name,
        score,
        date: timestamp,
      });

      history.push("/welcome");
    } catch (error) {
      throw new Error(error);
    }
  };

  const classes = useStyles();

  return (
    <>
      <Typography variant="h3" className={classes.scoreText}>
        You scored {score}!
      </Typography>

      <Button
        type="submit"
        color="primary"
        size="large"
        variant="contained"
        onClick={Save}
        className={classes.saveButton}
      >
        SAVE SCORE
      </Button>
    </>
  );
};
