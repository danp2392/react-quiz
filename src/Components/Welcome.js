import React from "react";
import { NavLink } from "react-router-dom";
import "../App.css";
import "../styling/Welcome.css";

export const Welcome = () => {
  return (
    <div className="container">
      <div className="background-img"></div>
      <div className="background">
        <NavLink to="/quiz" className="menu">
          Start Quiz
        </NavLink>
      </div>
      <div className="background">
        <NavLink to="/highscores" className="menu">
          High Scores
        </NavLink>
      </div>
    </div>
  );
};
