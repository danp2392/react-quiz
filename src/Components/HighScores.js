import React, { useEffect, useState } from "react";
import { db, scores } from "../auth/firebase";
import { useAuth } from "../auth/auth";
import "../App.css";
import "../styling/HighScores.css";
import { Navbar } from "./Navbar";
import firebase from "firebase";
import "firebase/auth";
import "firebase/database";

export const HighScores = () => {
  const [getUserData, setUserData] = useState([]);
  //nned to get user
  //need tp get scoes from database
  const { currentUser } = useAuth();

  /**function getData() {
    try {
      db.collection("scores")
        .where("name", "==", currentUser.displayName)
        .orderBy("date", "desc")
        .limit(10)
        .get()
        .then((snapshot) => {
          const userScoreData = [];
          snapshot.docs.forEach((doc) => {
            const data = doc.data();
            userScoreData.push(data);
          });

          setUserData(userScoreData);
        });
    } catch (e) {
      console.log(e);
    }
  } */

  const scoreDb = firebase.database();
  function dbListener() {
    db.collection("scores").onSnapshot((snap) => {
      const userScoreData = [];
      snap.docs.forEach((doc) => {
        const data = doc.data();
        userScoreData.push(data);
      });
      console.log(userScoreData);
    });
  }

  useEffect(() => {
    dbListener();
  });

  return (
    <>
      <Navbar />
      <div className="container">
        <div id="scores-div">
          <div id="user-div">Hello... {currentUser.displayName}</div>
          Your last 10 scores...
          <ul>
            {getUserData &&
              getUserData.map((item) => (
                <li key={item.date}>
                  <div className="score">{item.score} points</div>
                  <div id="seperator">-</div>
                  <div className="score">
                    {
                      //new Date(item.date.seconds * 1000).toLocaleDateString(
                      //"en-UK")
                    }
                  </div>
                </li>
              ))}
          </ul>
        </div>
      </div>
    </>
  );
};
