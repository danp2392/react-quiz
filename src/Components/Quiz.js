import React, { useEffect, useState } from "react";
import { LoadData } from "../api/api_connection";
import { decodeHTML } from "../functions/decodeHTML";
import { Question, ListAnswer } from "./Question";
import { useAuth } from "../auth/auth";
import { SaveScore } from "./SaveScore";
import { RandomiseAnswers } from "../functions/randomizeFunction";
import { Navbar } from "../Components/Navbar";

export const Quiz = () => {
  const [quizData, setQuizData] = useState([]);
  const [getQuestion, setQuestion] = useState([]);
  const [getCorrect, setCorrect] = useState("");
  const [getAnswers, setAnswers] = useState([]);
  const [getScore, setScore] = useState(0);
  const [currentQuestionId, setQuestionId] = useState(0);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const loadingData = async () => {
      const data = await LoadData();
      try {
        setQuizData(data);
        setIsLoading(false);
      } catch (e) {
        throw new Error(e);
      }
    };
    loadingData();
  }, []);

  const { currentUser } = useAuth();

  useEffect(() => {
    const runData = () => {
      setQuestion(quizData.map((item) => item.question));

      setAnswers(quizData.map((item) => RandomiseAnswers(item.answers)));

      setCorrect(quizData.map((item) => item.answer));
    };
    runData();
  }, [quizData]);

  function checkAnswer(e) {
    const getValue = e.target.innerHTML;
    setQuestionId(currentQuestionId + 1);
    if (getValue === getCorrect[currentQuestionId]) {
      setScore(getScore + 1);
    }
  }

  return isLoading ? (
    <div id="loading-div">LOADING...</div>
  ) : currentQuestionId < quizData.length ? (
    <>
      <Navbar />

      <Question
        id={currentQuestionId}
        question={decodeHTML(getQuestion[currentQuestionId])}
      />

      <ListAnswer
        id={currentQuestionId}
        answers={getAnswers[currentQuestionId]}
        checkAnswer={(answer) => checkAnswer(answer)}
      />
    </>
  ) : (
    <SaveScore score={getScore} name={currentUser.displayName} />
  );
};
