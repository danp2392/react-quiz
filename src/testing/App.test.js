import { render } from "@testing-library/react";
import Enzyme from "enzyme";
import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { Welcome } from "../Components/Welcome";
import { NavLink } from "react-router-dom";
import { shallow } from "enzyme";

Enzyme.configure({
  adapter: new Adapter(),
});

describe("welcome page", () => {
  const wrapper = shallow(<Welcome />);

  it("should render the welcome page", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should contain two nav links", () => {
    //find menu
    const navLink = wrapper.find("NavLink");
    expect(navLink).toHaveLength(2);
  });

  it("should contain a nav link called start quiz", () => {
    const navLinkDiv = wrapper.find("div");
    expect(
      navLinkDiv.containsMatchingElement(
        <NavLink to="/quiz">Start Quiz</NavLink>
      )
    ).toBe(true);
  });
  it("should contain a nav link called high scores", () => {
    const navLinkDiv = wrapper.find("div");
    expect(
      navLinkDiv.containsMatchingElement(
        <NavLink to="/highscores">High Scores</NavLink>
      )
    ).toBe(true);
  });
});
