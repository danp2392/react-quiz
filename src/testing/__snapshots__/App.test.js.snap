// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render a navbar page 1`] = `
ReactShallowRenderer {
  "_context": null,
  "_element": null,
  "_forcedUpdate": false,
  "_instance": null,
  "_newState": null,
  "_rendered": null,
  "_rendering": false,
  "_updater": Updater {
    "_callbacks": Array [],
    "_renderer": [Circular],
  },
}
`;

exports[`welcome page render welcome page 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`welcome page should render the welcome page 1`] = `
<div
  className="container"
>
  <div
    className="background-img"
  />
  <div
    className="background"
  >
    <[object Object]
      className="menu"
      to="/quiz"
    >
      Start Quiz
    </[object Object]>
  </div>
  <div
    className="background"
  >
    <[object Object]
      className="menu"
      to="/highscores"
    >
      High Scores
    </[object Object]>
  </div>
</div>
`;
